src/libraries/System.Private.CoreLib/src/System/Globalization/OrdinalCasing.Icu.cs

```diff
@@ -196,72 +197,90 @@ internal static int CompareStringIgnoreCase(ref char strA, int lengthA, ref char
             ref char charA = ref strA;
             ref char charB = ref strB;
 
-            while (length != 0)
+            int index = 0;
+
+            while (index < length)
             {
-                 // optimize for Ascii cases
-                if (charA <= '\u00FF' || length == 1 || !char.IsHighSurrogate(charA) || !char.IsHighSurrogate(charB))
+                char a = charA;
+                char b = charB;
+
+                if (!char.IsHighSurrogate(a) || index >= lengthA - 1 || !char.IsLowSurrogate(Unsafe.Add(ref charA, 1)))
```

Comment:
Couldn't charA + 1 also change while checking if a is a high surrogate?

Reply:
Yes. I have fixed it. thanks!